queasars/job_shop_scheduling/serialization.py

```diff
@@ -0,0 +1,183 @@
+    Machine,
+    Operation,
+    Job,
+    JobShopSchedulingProblemInstance,
+    ScheduledOperation,
+    JobShopSchedulingResult,
+)
+
+
+class JSSPJSONEncoder(JSONEncoder):
+    """
+    JSONEncoder class for encoding objects of the job shop scheduling data classes in
+    queasars.job_shop_scheduling.problem_instances
+    This class can serialize:
+        Machine,
+        Operation,
+        Job,
+        JobShopSchedulingProblemInstance,
+        ScheduledOperation,
+        JobShopSchedulingResult,
+    """
+
+    def default(self, o: Any) -> Any:
+
+        if isinstance(o, tuple):
+            return {"tuple": [self.default(entry) for entry in o]}
+
+        if isinstance(o, list):
+            return {"list": [self.default(entry) for entry in o]}
+
+        if isinstance(o, dict):
+            return {"dict": self.default(list(o.items()))}
+
+        if isinstance(o, Machine):
+            return {"machine_name": o.name}
+
+        if isinstance(o, Operation):
+            return {
+                "operation_name": o.name,
+                "operation_job_name": o.job_name,
+                "operation_machine": self.default(o.machine),
+                "operation_processing_duration": o.processing_duration,
+            }
+
+        if isinstance(o, Job):
+            return {"job_name": o.name, "job_operations": self.default(o.operations)}
+
+        if isinstance(o, JobShopSchedulingProblemInstance):
+            return {
+                "jssp_instance_name": o.name,
+                "jssp_instance_machines": self.default(o.machines),
+                "jssp_instance_jobs": self.default(o.jobs),
+            }
+
+        if isinstance(o, ScheduledOperation):
+            return {
+                "scheduled_operation": self.default(o.operation),
+                "scheduled_start_time": self.default(o.start),
+            }
+
+        if isinstance(o, JobShopSchedulingResult):
+            return {
+                "jssp_result_problem_instance": self.default(o.problem_instance),
+                "jssp_result_schedule": self.default(o.schedule),
+            }
+
+        return o
+
+
+class JSSPJSONDecoder(JSONDecoder):
+    """
+    JSONDecoder class for decoding objects of the job shop scheduling data classes in
+    queasars.job_shop_scheduling.problem_instances
+    This class can decode:
+        Machine,
+        Operation,
+        Job,
+        JobShopSchedulingProblemInstance,
+        ScheduledOperation,
+        JobShopSchedulingResult,
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, object_dict):
+
+        if "tuple" in object_dict:
+            return self.parse_tuple(object_dict=object_dict)
+
+        if "list" in object_dict:
+            return self.parse_list(object_dict=object_dict)
+
+        if "dict" in object_dict:
+            return self.parse_dict(object_dict=object_dict)
```

Comment:
This seems to me like it could shadow some errors, e.g.
```
{
  "dict": {},
  "list: []
}
```
would be parseble because the parser only deserializes the "list" object as this is tested for earlier. Or am I misunderstanding something?

Reply:
Right, this would be parseable, although the serialization would never create such an object. I added a check to check that in this case the key is the only entry in the object dictionary.

Reply:
Ah yes, that's a solution 👍